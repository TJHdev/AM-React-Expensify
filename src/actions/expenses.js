import uuid from 'uuid';
import database from '../firebase/firebase.js'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns function
// component dispatches object (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense: expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    
    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id: id
});

export const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).set(null).then(() => {
      dispatch(removeExpense({ id }));
    });
  }
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id: id,
  updates
});

export const startEditExpense = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    }); 
  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses: expenses
});

export const startSetExpenses = (expensesData = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
   
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
   
      dispatch(setExpenses(expenses));
    });
  }
};



 
