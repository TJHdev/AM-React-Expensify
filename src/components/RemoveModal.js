import React from 'react';
import Modal from 'react-modal';

const RemoveModal = (props) => (
  <Modal
    isOpen={Boolean(props.confirmRemoveExpense)}
    onRequestClose={props.handleCloseRemoveModal}
    contentLabel="Confirm Remove Expense"
    closeTimeoutMS={200}
    className="modal"
    ariaHideApp={false}
  >
    <h3 className="modal__title">Confirm Delete Expense</h3>
    <div className="modal__container">
      <button className="button" onClick={props.onRemove}>Delete</button>
      <button className="button test" onClick={props.handleCloseRemoveModal}>No</button>
    </div>
  </Modal>
);

export default RemoveModal;