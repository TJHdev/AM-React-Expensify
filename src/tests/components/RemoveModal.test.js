import React from 'react';
import { shallow } from 'enzyme';
import RemoveModal from '../../components/RemoveModal';
import expenses from '../fixtures/expenses';

let onRemove, handleCloseRemoveModal, expense, wrapper;

// TODO need help

beforeEach(() => {
  const state = {
    confirmRemoveExpense: null
  };
  history = { push: jest.fn() };
  expense = { id: '123' };
  handleCloseRemoveModal = jest.fn();
  onRemove = jest.fn();
  wrapper = shallow(
    <RemoveModal 
      confirmRemoveExpense={state.confirmRemoveExpense} 
      handleCloseRemoveModal={handleCloseRemoveModal}
      onRemove={onRemove}
    />
  );
});

test('should render RemoveModal properly', () => {
  expect(wrapper).toMatchSnapshot();    
});


test('should handle removeExpense', () => {
  const wrapper2 = shallow(<RemoveModal onRemove={onRemove} />);

  console.log(wrapper.find('button').at(1));
  wrapper.find('button').at(0).simulate('click');
  wrapper2.instance().onRemove();

  expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(handleCloseRemoveModal).toHaveBeenCalledTimes(1);
});
