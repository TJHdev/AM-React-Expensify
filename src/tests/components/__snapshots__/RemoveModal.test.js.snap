// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RemoveModal properly 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Confirm Remove Expense"
  isOpen={false}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3
    className="modal__title"
  >
    Confirm Delete Expense
  </h3>
  <div
    className="modal__container"
  >
    <button
      className="button"
      onClick={[MockFunction]}
    >
      Delete
    </button>
    <button
      className="button test"
      onClick={[MockFunction]}
    >
      No
    </button>
  </div>
</Modal>
`;
